Upgrading Control Plane fails due to invalid memory address

**What steps did you take and what happened:**
During upgrade, due to some unknown error, the pod crashes. 
Working on getting the cause.

Note: The main focus of this bug report is on fixing the nil pointer issue, much like  #2613.

**What did you expect to happen:**
A graceful failure

**Environment:**

- Cluster-api version: v1alpha3
- Minikube/KIND version:
- Kubernetes version: v1.17.3
- Ubuntu 18.04.4

/kind bug

```
kubectl logs capi-kubeadm-control-plane-controller-manager-56d8897c7d-cnn9p -n capi-kubeadm-control-plane-system -c manager
I0311 09:01:16.374402       1 listener.go:40] controller-runtime/metrics "msg"="metrics server is starting to listen"  "addr"="127.0.0.1:8080"
I0311 09:01:16.376447       1 main.go:119] setup "msg"="starting manager"  
I0311 09:01:16.378539       1 leaderelection.go:242] attempting to acquire leader lease  capi-kubeadm-control-plane-system/kubeadm-control-plane-manager-leader-election-capi...
I0311 09:01:16.382102       1 internal.go:356] controller-runtime/manager "msg"="starting metrics server"  "path"="/metrics"
I0311 09:01:33.837344       1 leaderelection.go:252] successfully acquired lease capi-kubeadm-control-plane-system/kubeadm-control-plane-manager-leader-election-capi
I0311 09:01:33.847483       1 controller.go:164] controller-runtime/controller "msg"="Starting EventSource"  "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"version":"","infrastructureTemplate":{},"kubeadmConfigSpec":{}},"status":{"initialized":false,"ready":false}}}
I0311 09:01:33.959149       1 controller.go:164] controller-runtime/controller "msg"="Starting EventSource"  "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"clusterName":"","bootstrap":{},"infrastructureRef":{}},"status":{"bootstrapReady":false,"infrastructureReady":false}}}
I0311 09:01:34.073241       1 controller.go:164] controller-runtime/controller "msg"="Starting EventSource"  "controller"="kubeadmcontrolplane" "source"={"Type":{"metadata":{"creationTimestamp":null},"spec":{"controlPlaneEndpoint":{"host":"","port":0}},"status":{"infrastructureReady":false,"controlPlaneInitialized":false}}}
I0311 09:01:34.228410       1 controller.go:171] controller-runtime/controller "msg"="Starting Controller"  "controller"="kubeadmcontrolplane"
I0311 09:01:34.228487       1 controller.go:190] controller-runtime/controller "msg"="Starting workers"  "controller"="kubeadmcontrolplane" "worker count"=10
I0311 09:01:34.359547       1 kubeadm_control_plane_controller.go:252] controllers/KubeadmControlPlane "msg"="Upgrading Control Plane" "cluster"="test1" "kubeadmControlPlane"="test1-controlplane" "namespace"="metal3" 
E0311 09:01:35.049926       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 278 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x16a8280, 0x26f9f60)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:48 +0x82
panic(0x16a8280, 0x26f9f60)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).upgradeControlPlane(0xc0003aa120, 0x1b02b40, 0xc00009e050, 0xc000241b00, 0xc00020a500, 0xc0003441b0, 0xc000344240, 0xc0000a0c88, 0x1, 0x1, ...)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:376 +0x5ff
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).reconcile(0xc0003aa120, 0x1b02b40, 0xc00009e050, 0xc000241b00, 0xc00020a500, 0x0, 0x0, 0x0, 0xc00042ca1a)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:253 +0x1250
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).Reconcile(0xc0003aa120, 0xc00042ca1a, 0x6, 0xc00058ed00, 0x12, 0xc0003eec00, 0x0, 0x0, 0x0)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:190 +0x6a3
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001ca480, 0x1709460, 0xc0003f7c20, 0x440500)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001ca480, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001ca480)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0003fcb70)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0003fcb70, 0x3b9aca00, 0x0, 0x1, 0xc000042a20)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc0003fcb70, 0x3b9aca00, 0xc000042a20)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:193 +0x328
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x152ca2f]

goroutine 278 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/runtime/runtime.go:55 +0x105
panic(0x16a8280, 0x26f9f60)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).upgradeControlPlane(0xc0003aa120, 0x1b02b40, 0xc00009e050, 0xc000241b00, 0xc00020a500, 0xc0003441b0, 0xc000344240, 0xc0000a0c88, 0x1, 0x1, ...)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:376 +0x5ff
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).reconcile(0xc0003aa120, 0x1b02b40, 0xc00009e050, 0xc000241b00, 0xc00020a500, 0x0, 0x0, 0x0, 0xc00042ca1a)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:253 +0x1250
sigs.k8s.io/cluster-api/controlplane/kubeadm/controllers.(*KubeadmControlPlaneReconciler).Reconcile(0xc0003aa120, 0xc00042ca1a, 0x6, 0xc00058ed00, 0x12, 0xc0003eec00, 0x0, 0x0, 0x0)
	/workspace/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:190 +0x6a3
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001ca480, 0x1709460, 0xc0003f7c20, 0x440500)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001ca480, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001ca480)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0003fcb70)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0003fcb70, 0x3b9aca00, 0x0, 0x1, 0xc000042a20)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc0003fcb70, 0x3b9aca00, 0xc000042a20)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.2/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.0/pkg/internal/controller/controller.go:193 +0x328

```

--- a/controlplane/kubeadm/controllers/fakes_test.go
+++ b/controlplane/kubeadm/controllers/fakes_test.go
@@ -20,6 +20,7 @@ import (
 	"context"
 	"errors"
 
+	"github.com/blang/semver"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
 	"sigs.k8s.io/cluster-api/controlplane/kubeadm/internal"
 	"sigs.k8s.io/controller-runtime/pkg/client"
@@ -70,6 +71,26 @@ func (f fakeWorkloadCluster) ForwardEtcdLeadership(_ context.Context, _ *cluster
 
 func (f fakeWorkloadCluster) ClusterStatus(_ context.Context) (internal.ClusterStatus, error) {
 	return f.Status, nil
+}
+
+func (f fakeWorkloadCluster) ReconcileKubeletRBACRole(ctx context.Context, version semver.Version) error {
+	return nil
+}
+
+func (f fakeWorkloadCluster) ReconcileKubeletRBACBinding(ctx context.Context, version semver.Version) error {
+	return nil
+}
+
+func (f fakeWorkloadCluster) UpdateKubernetesVersionInKubeadmConfigMap(ctx context.Context, version semver.Version) error {
+	return nil
+}
+
+func (f fakeWorkloadCluster) UpdateEtcdVersionInKubeadmConfigMap(ctx context.Context, imageRepository, imageTag string) error {
+	return nil
+}
+
+func (f fakeWorkloadCluster) UpdateKubeletConfigMap(ctx context.Context, version semver.Version) error {
+	return nil
 }
 
 type fakeMigrator struct {

--- a/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go
+++ b/controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go
@@ -373,7 +373,7 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context,
 		return ctrl.Result{}, errors.Wrap(err, "failed to update the kubernetes version in the kubeadm config map")
 	}
 
-	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.Local != nil {
+	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration != nil && kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.Local != nil {
 		meta := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.Local.ImageMeta
 		if err := workloadCluster.UpdateEtcdVersionInKubeadmConfigMap(ctx, meta.ImageRepository, meta.ImageTag); err != nil {
 			return ctrl.Result{}, errors.Wrap(err, "failed to update the etcd version in the kubeadm config map")

--- a/controlplane/kubeadm/controllers/kubeadm_control_plane_controller_test.go
+++ b/controlplane/kubeadm/controllers/kubeadm_control_plane_controller_test.go
@@ -2130,3 +2130,42 @@ func TestKubeadmControlPlaneReconciler_failureDomainForScaleUp(t *testing.T) {
 		})
 	}
 }
+
+func TestKubeadmControlPlaneReconciler_upgradeControlPlane(t *testing.T) {
+	g := NewWithT(t)
+
+	cluster, kcp, genericMachineTemplate := createClusterWithControlPlane()
+	kcp.Spec.Version = "v1.17.3"
+	kcp.Spec.KubeadmConfigSpec.ClusterConfiguration = nil
+
+	fakeClient := newFakeClient(g, cluster.DeepCopy(), kcp.DeepCopy(), genericMachineTemplate.DeepCopy())
+
+	r := &KubeadmControlPlaneReconciler{
+		Client:   fakeClient,
+		Log:      log.Log,
+		recorder: record.NewFakeRecorder(32),
+		managementCluster: &fakeManagementCluster{
+			Management: &internal.Management{Client: fakeClient},
+			Workload:   fakeWorkloadCluster{},
+		},
+	}
+
+	result, err := r.initializeControlPlane(context.Background(), cluster, kcp)
+	g.Expect(result).To(Equal(ctrl.Result{Requeue: true}))
+	g.Expect(err).NotTo(HaveOccurred())
+
+	machineList := &clusterv1.MachineList{}
+	g.Expect(fakeClient.List(context.Background(), machineList, client.InNamespace(cluster.Namespace))).To(Succeed())
+	g.Expect(machineList.Items).NotTo(BeEmpty())
+	g.Expect(machineList.Items).To(HaveLen(1))
+
+	machineCollection := internal.NewFilterableMachineCollection(&machineList.Items[0])
+	result, err = r.upgradeControlPlane(context.Background(), cluster, kcp, machineCollection, machineCollection)
+
+	g.Expect(machineList.Items[0].Annotations).To(HaveKey(controlplanev1.SelectedForUpgradeAnnotation))
+
+	// TODO flesh out the rest of this test - this is currently least-effort to confirm a fix for an NPE when updating
+	// the etcd version
+	g.Expect(result).To(Equal(ctrl.Result{}))
+	g.Expect(err).To(Equal(&capierrors.RequeueAfterError{RequeueAfter: HealthCheckFailedRequeueAfter}))
+}

